examples/experimental/dagster-airlift/dagster_airlift/core/sensor/builder.py

```diff
@@ -190,3 +194,70 @@ def materializations_and_requests_from_batch_iter(
             if mats
             else None
         )
+
+
+def build_synthetic_asset_materializations(
+    context: SensorEvaluationContext,
+    airflow_instance: AirflowInstance,
+    dag_run: DagRun,
+    airflow_data: AirflowDefinitionsData,
+) -> List[AssetMaterialization]:
+    """In this function we need to return the asset materializations we want to synthesize
+    on behalf of the user.
+
+    This happens when the user has modeled an external asset that has a corresponding
+    task in airflow which is not proxied to Dagster. We want to detect the case
+    where there is a successful airflow task instance that is mapped to a
+    dagster asset but is _not_ proxied, and then synthensize a materialization
+    for observability.
+
+    We do this by querying for successful task instances in Airflow. And then
+    for each successful task we see it there exists a Dagster Run tagged with
+    the run id. If there is not Dagster run, we know the task was not proxied.
+
+    Task instances are mutable in Airflow, so we are not guaranteed to register
+    every task instance. If, for example, the sensor is paused, and then there are
+    multiple task clearings, we will only register the last materialization.
+
+    This also currently does not support dynamic tasks in Airflow, in which case
+    the use should instead map at the dag-level granularity.
+    """
+    task_instances = airflow_instance.get_task_instance_batch(
+        run_id=dag_run.run_id,
+        dag_id=dag_run.dag_id,
+        task_ids=[
+            task_id for task_id in airflow_data.serialized_data.task_ids_in_dag(dag_run.dag_id)
+        ],
+        states=["success"],
+    )
+
+    check.invariant(
+        len({ti.task_id for ti in task_instances}) == len(task_instances),
+        "Assuming one task instance per task_id for now. Dynamic Airflow tasks not supported.",
+    )
+
+    # TODO: Need to properly paginate this.
+    dagster_runs = context.instance.get_runs(
+        filters=RunsFilter(tags={DAG_RUN_ID_TAG_KEY: dag_run.run_id}),
+        limit=1000,
```

Comment:
looks paginated to me 😎

Reply:
Adding https://linear.app/dagster-labs/issue/FOU-444/make-sensor-work-with-an-airflow-dag-run-that-has-more-than-1000 to track